### PR TITLE
Implement resource binding type prefix mismatch errors

### LINKED ISSUE #57886: [HLSL] Validate register binding type match the decl type
Report error for case like RWBuffer<float> Uav : register(t3, space1);
The register should be u3 for Uav.

AC:
Error and diagnose if you try to bind a resource to an incompatible register space
Add something to classify the resource types

### PR DESCRIPTION
**Pull request by bob80905**
There are currently no diagnostics being emitted for when a resource is bound to a register with an incorrect binding type prefix. For example, a CBuffer type resource should be bound with a a binding type prefix of 'b', but if instead the prefix is 'u', no errors will be emitted. This PR implements a set of diagnostics for when the prefix is not expected.
Fixes #57886

### CODE REVIEW

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 1]
```diff
@@ -7367,6 +7367,75 @@ static void handleHLSLResourceBindingAttr(Sema &S, Decl *D,
...
+      break;
+    }
+    case llvm::hlsl::ResourceClass::Invalid: {
+      llvm_unreachable("Resource class should be valid.");
+      break;
+    }
+
+    default:
+      break;
+    }
```
Comment (Keenuts) [inline clang/lib/Sema/SemaDeclAttr.cpp:7436]: Shall the `Invalid` case be merged with the default case? So if an new enum value is added one day, we get a real crash vs a silent pass?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 2]
```diff
@@ -7367,6 +7367,75 @@ static void handleHLSLResourceBindingAttr(Sema &S, Decl *D,
...
+  }
+
   // FIXME: check reg type match decl. Issue
```
Comment (python3kgae) [inline clang/lib/Sema/SemaDeclAttr.cpp:7439]: This FIXME could be removed.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 3]
```diff
@@ -7367,6 +7367,75 @@ static void handleHLSLResourceBindingAttr(Sema &S, Decl *D,
...
+        RC = llvm::hlsl::ResourceClass::CBuffer;
+        varTy = "cbuffer";
+      } else {
+        RC = llvm::hlsl::ResourceClass::CBuffer;
+        varTy = "tbuffer";
+      }
+    }
+    switch (RC) {
+    case llvm::hlsl::ResourceClass::SRV: {
+      if (Slot.substr(0, 1) != "t")
```
Comment (python3kgae) [inline clang/lib/Sema/SemaDeclAttr.cpp:7406]: Shall we add a char regName = Slot[0] to replace Slot.substr(0,1).
And maybe need to check !Slot.empty().

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 4]
```diff
@@ -0,0 +1,65 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -o - -fsyntax-only %s -verify
+
+
+// expected-error@+1 {{invalid register name prefix 'b' for register type 'RWBuffer' (expected 'u')}}
+RWBuffer<int> a : register(b2, space1);
```
Comment (python3kgae) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:5]: Could we add a test for RWBuffer<int> b: register();/RWBuffer<int> b: register(""); to make sure empty string works?

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 1 above)
Comment (llvm-beanz) [reply clang/lib/Sema/SemaDeclAttr.cpp:7436]: Or if we're covering all cases in the switch we should leave off the default case so that it produces a warning if we add a new resource class.

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 3 above)
Comment (llvm-beanz) [reply clang/lib/Sema/SemaDeclAttr.cpp:7406]: Or we could replace all the `Slot.substr(0,1)` calls just with `Slot[0]` and compare characters instead of strings.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 5]
```diff
@@ -7403,41 +7403,36 @@ static void handleHLSLResourceBindingAttr(Sema &S, Decl *D,
...
     }
     case llvm::hlsl::ResourceClass::Sampler: {
-      if (Slot.substr(0, 1) != "s")
+      if (Slot[0] != 's')
         S.Diag(ArgLoc, diag::err_hlsl_mismatching_register_type_and_name)
             << Slot.substr(0, 1) << varTy << "s";
       break;
     }
-    case llvm::hlsl::ResourceClass::Invalid: {
+    default: {
```
Comment (llvm-beanz) [inline clang/lib/Sema/SemaDeclAttr.cpp:7429]: It would be better to have this be the Invalid case. Then the compiler will warn if new enum values get added that aren't handled.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 6]
```diff
@@ -119,8 +119,10 @@ struct BuiltinTypeDeclBuilder {
                                                 ResourceKind RK, bool IsROV) {
     if (Record->isCompleteDefinition())
       return *this;
-    Record->addAttr(HLSLResourceAttr::CreateImplicit(Record->getASTContext(),
-                                                     RC, RK, IsROV));
+    HLSLResourceAttr *attr = HLSLResourceAttr::CreateImplicit(
+        Record->getASTContext(), RC, RK, IsROV);
+
+    Record->addAttr(attr);
```
[the remark this answers is not in the document]
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:125]: Correct, just a result of some debugging, I'll revert it.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 7]
```diff
@@ -7337,8 +7337,8 @@ static void handleHLSLResourceBindingAttr(Sema &S, Decl *D,
   // Validate.
   if (!Slot.empty()) {
     switch (Slot[0]) {
-    case 'u':
     case 'b':
+    case 'u':
```
Comment (farzonl) [inline clang/lib/Sema/SemaDeclAttr.cpp:7341]: Why make this change? if its for alphabetical reasons, stu should be the order, if not why leave as is?

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 8]
```diff
@@ -482,15 +484,18 @@ static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
...
-  Decl = BuiltinTypeDeclBuilder(*SemaPtr, HLSLNamespace, "RWBuffer")
-             .addSimpleTemplateParams({"element_type"})
-             .Record;
+  Decl =
+      BuiltinTypeDeclBuilder(*SemaPtr, HLSLNamespace, "RWBuffer")
+          .addSimpleTemplateParams({"element_type"})
+          .annotateResourceClass(ResourceClass::UAV, ResourceKind::TypedBuffer,
+                                 /*IsROV=*/false)
+          .Record;
+
```
[the remark this answers is not in the document]
Comment (bob80905) [reply clang/lib/Sema/HLSLExternalSemaSource.cpp:496]: We do need to move it, since the attribute won't be available when it is time to diagnose otherwise. The extra parameters have been removed.

File: clang/lib/Sema/HLSLExternalSemaSource.cpp [hunk 9]
```diff
@@ -478,33 +478,35 @@ void HLSLExternalSemaSource::defineTrivialHLSLTypes() {
...
 static BuiltinTypeDeclBuilder setupBufferType(CXXRecordDecl *Decl, Sema &S,
-                                              ResourceClass RC, ResourceKind RK,
-                                              bool IsROV) {
+                                              ResourceClass RC) {
   return BuiltinTypeDeclBuilder(Decl)
       .addHandleMember()
-      .addDefaultHandleConstructor(S, RC)
-      .annotateResourceClass(RC, RK, IsROV);
+      .addDefaultHandleConstructor(S, RC);
 }
```
Comment (bogner) [inline clang/lib/Sema/HLSLExternalSemaSource.cpp:485]: Let's either remove this function (since it isn't buying us all that much at this point) or rename it to be specifically about setting up the buffer's handle (since that's all it does).

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 10]
```diff
@@ -7334,6 +7334,81 @@ static void handleHLSLShaderAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+  HLSLBufferDecl *CBufferOrTBuffer = dyn_cast<HLSLBufferDecl>(D);
+  if (!SamplerUAVOrSRV && !CBufferOrTBuffer)
+    return;
+
+  llvm::hlsl::ResourceClass DeclResourceClass;
+  StringRef varTy = "";
+  if (SamplerUAVOrSRV) {
+    const Type *Ty = SamplerUAVOrSRV->getType()->getPointeeOrArrayElementType();
+    if (!Ty)
+      llvm_unreachable("Resource class should have an element type.");
```
Comment (bogner) [inline clang/lib/Sema/SemaDeclAttr.cpp:7351]: I probably just read too many RFCs, but it'd be better to use "must" rather than "should" in this message.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 11]
```diff
@@ -7334,6 +7334,81 @@ static void handleHLSLShaderAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+                                        Decl *D, StringRef &Slot) {
+  // Samplers, UAVs, and SRVs are VarDecl types
+  VarDecl *SamplerUAVOrSRV = dyn_cast<VarDecl>(D);
+  // Cbuffers and Tbuffers are HLSLBufferDecl types
+  HLSLBufferDecl *CBufferOrTBuffer = dyn_cast<HLSLBufferDecl>(D);
+  if (!SamplerUAVOrSRV && !CBufferOrTBuffer)
+    return;
+
+  llvm::hlsl::ResourceClass DeclResourceClass;
+  StringRef varTy = "";
```
Comment (bogner) [inline clang/lib/Sema/SemaDeclAttr.cpp:7347]: LLVM style prefers `VarTy`

File: clang/test/SemaHLSL/resource_binding_attr_error.hlsl [hunk 12]
```diff
@@ -45,7 +45,7 @@ void foo2() {
   extern RWBuffer<float> U2 : register(u5);
 }
 // FIXME: expect-error once fix https://github.com/llvm/llvm-project/issues/57886.
-float b : register(u0, space1);
+// float b : register(u0, space1) {}
```
Comment (bogner) [inline clang/test/SemaHLSL/resource_binding_attr_error.hlsl:48]: Shouldn't we be adding the `// expected-error:` here rather than commenting this out?

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 13]
```diff
@@ -0,0 +1,74 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -o - -fsyntax-only %s -verify
+
+// the below will cause an llvm unreachable, because RWBuffers don't have resource attributes yet
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'b' for register type 'RWBuffer' (expected 'u')}}
+// NOT YET IMPLEMENTED RWBuffer<int> a : register(b2, space1);
```
Comment (bogner) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:5]: This seems incorrect - RWBuffer is the one resource type that's actually available

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+// expected-error@+1 {{invalid register name prefix 's' for register type 'cbuffer' (expected 'b')}}
+cbuffer f : register(s2, space1) {}
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 't' for register type 'Sampler' (expected 's')}}
+// Can this type just be Sampler instead of SamplerState?
+// NOT YET IMPLEMENTED SamplerState MySampler : register(t3, space1);
+
+// expected-error@+1 {{invalid register name prefix 's' for register type 'tbuffer' (expected 'b')}}
+tbuffer f : register(s2, space1) {}
```
Comment (tex3d) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:58]: As was stated in an earlier (resolved but seemingly not fixed) comment, tbuffer must be bound to `t`.

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+// NOT YET IMPLEMENTED Texture2D<float> Texture : register(s0);
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'u' for register type 'Texture2DMS' (expected 't')}}
+// NOT YET IMPLEMENTED Texture2DMS<float4, 4> T2DMS_t2 : register(u2)
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 't' for register type 'RWTexture3D' (expected 'u')}}
+// NOT YET IMPLEMENTED RWTexture3D<float4> RWT3D_u1 : register(t1)
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'b' for register type 'Texture2DMS' (expected 't' or 's')}}
+// NOT YET IMPLEMENTED TextureCube TCube_b2 : register(B2);
```
Comment (tex3d) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:24]: 's' should not be considered a valid register binding for TextureCube, and the message says 'Texture2DMS' instead of 'TextureCube'.

Comment (tex3d) [reply clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:24]: Same with all other `(expected 't' or 's')` below.

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+// NOT YET IMPLEMENTED TextureCubeArray TCubeArray_t2 : register(b2);
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'b' for register type 'Texture2DMS' (expected 't' or 's')}}
+// NOT YET IMPLEMENTED Texture1DArray T1DArray_t2 : register(b2);
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'u' for register type 'Texture2DMS' (expected 't' or 's')}}
+// NOT YET IMPLEMENTED Texture2DArray T2DArray_b2 : register(B2);
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'u' for register type 'Texture2DMS' (expected 'b' or 'c' or 'i')}}
+// NOT YET IMPLEMENTED Texture2DMSArray<float4> msTextureArray : register(t2, space2);
```
Comment (tex3d) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:36]: Expected would be 't', not `'b' or 'c' or 'i'`.  In fact, that group ('b', 'c', 'i') were legacy constant register bindings for DX9, which is where I suspect this comes from, where 'b' meant a special bool constant and 'i' mapped to special loop constant values (used in DX9 shader models), and 'c' was a float constant value.  'b' was since used for constant buffer binding location as well, so there is a bit of a collision, but not on the same variable type, since the 'b', 'c', 'i' bindings were on global numeric variables, not on resource objects or cbuffer/tbuffer declarations.

So, we no longer support the DX9 'b','c','i' numeric constant bindings the same way, but we do reuse 'c' for manual cbuffer layout, and 'b' for cbuffer or ConstantBuffer binding.

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl (hunk 14 above)
Comment (bob80905) [reply clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl]: My bad, resolved the comment before pushing the commit, I'll try and remember to swap that order next time.

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl (hunk 16 above)
Comment (bob80905) [reply clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:36]: `float b : register(u0, space1);`
How about in this case? I see evidence in past hlsl shaders that the b, c, or i diagnostic would be emitted here, would we expect 't' here instead?

Comment (bob80905) [reply clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl]: I've changed the expected binding prefix to 't' in the case above, working with that assumption now.

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 17]
```diff
@@ -7334,6 +7334,100 @@ static void handleHLSLShaderAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+      VarTy = "tbuffer";
+    }
+  }
+  switch (DeclResourceClass) {
+  case llvm::hlsl::ResourceClass::SRV: {
+    if (Slot[0] != 't')
+      S.Diag(ArgLoc, diag::err_hlsl_mismatching_register_resource_type_and_name)
+          << Slot.substr(0, 1) << VarTy
+          << (unsigned)llvm::hlsl::ResourceClass::SRV;
+    break;
```
Comment (bogner) [inline clang/lib/Sema/SemaDeclAttr.cpp:7394]: This code is all a little repetitive. Should we reverse the logic and exit early in the valid cases instead, and emit the diagnostic in the fallthrough?

```c++
  switch (DeclResourceClass) {
  case llvm::hlsl::ResourceClass::SRV:
    if (Slot[0] == 't')
      return;
  case llvm::hlsl::ResourceClass::UAV:
    if (Slot[0] == 'u')
      return;
  // ...
  }
  S.Diag(ArgLoc, diag::err_hlsl_mismatching_register_resource_type_and_name)
      << Slot.substr(0, 1) << VarTy
      << (unsigned)DeclResourceClass;
```

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 18]
```diff
@@ -0,0 +1,74 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -o - -fsyntax-only %s -verify
+
+// the below will cause an llvm unreachable, because RWBuffers don't have resource attributes yet
+// expected-error@+1  {{invalid register name prefix 'b' for register resource type 'RWBuffer' (expected 'u')}}
+RWBuffer<int> a : register(b2, space1);
+
+// the below will cause an llvm unreachable, because RWBuffers don't have resource attributes yet
```
Comment (bogner) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:7]: This comment is still incorrect.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 19]
```diff
@@ -12154,6 +12154,8 @@ def err_hlsl_missing_semantic_annotation : Error<
 def err_hlsl_init_priority_unsupported : Error<
   "initializer priorities are not supported in HLSL">;
 
+def err_hlsl_mismatching_register_resource_type_and_name: Error<"invalid register name prefix '%0' for register resource type '%1' (expected %select{'t'|'u'|'b'|'t'|'s'}2)">;
```
Comment (python3kgae) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:12157]: 't' happened twice.

Comment (bob80905) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:12157]: We need 't' twice, because both TBuffer and SRV Resource classes expect a 't' binding prefix. The diagnostic engine takes in the resource class enum as an integer value to select which binding prefix character to expect in its message, so we need 0 ( the enum val for ResourceClass::SRV) and 3 (the enum val for ResourceClass::TBuffer) to both map to 't'.

File: llvm/include/llvm/Frontend/HLSL/HLSLResource.h [hunk 20]
```diff
@@ -25,6 +25,7 @@ enum class ResourceClass : uint8_t {
   SRV = 0,
   UAV,
   CBuffer,
+  TBuffer,
```
Comment (python3kgae) [inline llvm/include/llvm/Frontend/HLSL/HLSLResource.h:28]: TBuffer should be a resource kind, not resource class.

Comment (bogner) [reply llvm/include/llvm/Frontend/HLSL/HLSLResource.h:28]: I hadn't noticed this. Xiang's right, this isn't correct. See `CGHLSLRuntime::finishCodeGen` - a tbuffer is annotated with a resource class of `SRV` as you would expect.

File: clang/test/SemaHLSL/resource_binding_attr_error.hlsl [hunk 21]
```diff
@@ -44,7 +44,7 @@ void foo2() {
   // expected-warning@+1 {{'register' attribute only applies to cbuffer/tbuffer and external global variables}}
   extern RWBuffer<float> U2 : register(u5);
 }
-// FIXME: expect-error once fix https://github.com/llvm/llvm-project/issues/57886.
+// expected-error@+1 {{invalid register name prefix 'u' for 'float' (expected 't')}}
```
Comment (python3kgae) [inline clang/test/SemaHLSL/resource_binding_attr_error.hlsl:47]: Why a float expected 't'?

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 22]
```diff
@@ -7334,6 +7334,92 @@ static void handleHLSLShaderAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
...
+    if (Slot[0] == 'u')
+      return;
+    break;
+  }
+  case llvm::hlsl::ResourceClass::CBuffer: {
+    // could be CBuffer or TBuffer
+    if (CBufferOrTBuffer->isCBuffer()) {
+      VarTy = "cbuffer";
+      if (Slot[0] == 'b')
+        return;
```
Comment (bogner) [inline clang/lib/Sema/SemaDeclAttr.cpp:7398]: This will likely crash once we add the [ConstantBuffer](https://learn.microsoft.com/en-us/windows/win32/direct3d12/resource-binding-in-hlsl#constant-buffers) template syntax, since we'll be able to have:
```hlsl
struct S {
  float4 g_MaxThreadIter;
  float4 g_Window;
};
ConstantBuffer<S> cb0 : register(b0);
```

We should at the very least put a TODO here, but maybe if we can drive cbuffer/tbuffer from attributes we could just do the right thing already?

File: clang/test/SemaHLSL/resource_binding_attr_error.hlsl (hunk 21 above)
Comment (bob80905) [reply clang/test/SemaHLSL/resource_binding_attr_error.hlsl:47]: I don't have a prior example to base this off of at hand, but I think I saw some hlsl with a builtin type bound to a register and the valid prefix was 't'.
To answer your question, the assignment is represented as a VarDecl, so the compiler assumes this resource is a Sampler, UAV, or  SRV. Then the resource type is analyzed to see if it's a builtin (float happens to be a builtin).
Then we check the binding prefix to see if it's 't', and if not, emit an error. float is a builtin type, and the prefix isn't t, so that's why this test case expects 't'.
What prefix should it expect, if any?

Comment (python3kgae) [reply clang/test/SemaHLSL/resource_binding_attr_error.hlsl:47]: dxc will report
error: invalid register specification, expected 'b', 'c', or 'i' binding
float t : register(u0);
          ^
fxc will just ignore the register(u0).

Comment (python3kgae) [reply clang/test/SemaHLSL/resource_binding_attr_error.hlsl:47]: Since here the float is in global scope, it will be treated as constant inside Global CBuffer.

Comment (bob80905) [reply clang/test/SemaHLSL/resource_binding_attr_error.hlsl:47]: According to @tex3d 's comment, the ` expected 'b', 'c', or 'i' binding` is no longer supported, so I don't think I should weave that into this implementation. I did ask what we should expect the prefix to be, but I haven't received an answer yet. Sounds to me like you think it should be 'b', because it's a constant inside a CBuffer?

Comment (tex3d) [reply clang/test/SemaHLSL/resource_binding_attr_error.hlsl:47]: We should consider deprecating the register binding of this style: `float b : register(c0);`, since it's not supported by DXC, and was only supported by FXC for DX9 targets.  Register binding should only be applicable to global resource/sampler declarations and cbuffer/tbuffer declarations.  So the answer to "what prefix should that be" is: we don't support register bindings on values that go into the constant buffer, so there is no valid prefix to use here.  The closest equivalent to these legacy register bindings that we do support are `packoffset` annotations.

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 23]
```diff
@@ -0,0 +1,73 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -o - -fsyntax-only %s -verify
+
+
+// expected-error@+1  {{invalid register name prefix 'b' for register resource type 'RWBuffer' (expected 'u')}}
+RWBuffer<int> a : register(b2, space1);
```
Comment (damyanp) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:5]: > float b : register(u0, space1);

This would also be invalid even if the register type were correct because you can't have spaces on global constants.

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+// NOT YET IMPLEMENTED TextureCube <float>  t8 : register(b8);
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'b' for register type 'TextureCubeArray' (expected 't')}}
+// NOT YET IMPLEMENTED TextureCubeArray TCubeArray_t2 : register(b2);
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'b' for register type 'Texture1DArray' (expected 't')}}
+// NOT YET IMPLEMENTED Texture1DArray T1DArray_t2 : register(b2);
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'u' for register type 'Texture2DArray' (expected 't')}}
+// NOT YET IMPLEMENTED Texture2DArray T2DArray_b2 : register(B2);
```
Comment (damyanp) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:32]: Should probably be one of:

```suggestion
// NOT YET IMPLEMENTED : {{invalid register name prefix 'u' for register type 'Texture2DArray' (expected 't')}}
// NOT YET IMPLEMENTED Texture2DArray T2DArray_b2 : register(u2);
```

```suggestion
// NOT YET IMPLEMENTED : {{invalid register name prefix 'B' for register type 'Texture2DArray' (expected 't')}}
// NOT YET IMPLEMENTED Texture2DArray T2DArray_b2 : register(B2);

```

File: clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+
+// NOT YET IMPLEMENTED : {{invalid register name prefix 'u' for register type 'Texture2DMSArray' (expected 't')}}
+// NOT YET IMPLEMENTED Texture2DMSArray<float4> msTextureArray : register(t2, space2);
```
Comment (damyanp) [inline clang/test/SemaHLSL/resource_binding_attr_error_mismatch.hlsl:35]: Error message also doesn't match the register in this case.

File: clang/test/AST/HLSL/resource_binding_attr.hlsl [hunk 26]
```diff
@@ -1,13 +1,15 @@
 // RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -ast-dump -o - %s | FileCheck %s
```
Comment (python3kgae) [inline clang/test/AST/HLSL/resource_binding_attr.hlsl:2]: Should we support
```
 float c : register(c10);
```
It is legal in dxc.

Comment (damyanp) [reply clang/test/AST/HLSL/resource_binding_attr.hlsl:2]: yes

**Comment by bob80905** [comment]
This PR was based on code from April 3, and the branch is far too stale to update at this point. I'm closing this, and have moved the implementation over to this PR:
https://github.com/llvm/llvm-project/pull/97103
